generate_workspace/src/test/java/com/google/devtools/build/workspace/maven/VersionResolverTest.java

```diff
@@ -0,0 +1,108 @@
+import org.junit.runners.JUnit4;
+import org.mockito.Mockito;
+
+/**
+ * Tests for {@link VersionResolver}.
+ */
+//TODO(petros): Test this actually selects highest version. VersionRangeResult is final class so Mockito cannot mock it.
+@RunWith(JUnit4.class)
+public class VersionResolverTest {
+
+  /**
+   * Ensures that an exception is thrown if there is a version range resolution exception.
+   * This occurs when aether is unable to resolve the version.
+   */
+  @Test
+  public void failsOnResolutionException() {
+    RepositorySystem system = Mockito.mock(RepositorySystem.class);
+    try {
+      Mockito.when(system.resolveVersionRange(anySession(), anyRangeRequest()))
+          .thenThrow(new VersionRangeResolutionException(any()));
+      Aether aether = Aether.builder().systemSession(anySession(), system).build();
+      VersionResolver resolver = new VersionResolver(aether);
+
+      resolver.resolveVersion("something", "something", "1.0");
+      fail();
+    } catch (VersionRangeResolutionException e) {
+      // This should be caught by the VersionResolver
+      fail();
+    } catch (ArtifactBuilder.InvalidArtifactCoordinateException e) {
+      String expected =
+          "Unable to find a version for something:something:1.0 due to Failed to resolve version range";
+      assertThat(e.getMessage()).isEqualTo(expected);
+    }
+  }
+
+  /**
+   * Ensures that an exception is thrown if there is an invalid version range. An invalid version range
+   * is one which is either (1) equal to null or (2) returns null when asked for highest version.
+   */
+  @Test
+  public void failsOnInvalidVersionRange() {
+    RepositorySystem system = Mockito.mock(RepositorySystem.class);
+
+    try {
+      // Using `anyRangeResult()` will ensure that rangeResult.highestVersion() == null.
+      Mockito.when(system.resolveVersionRange(anySession(), anyRangeRequest())).thenReturn(anyRangeResult());
+      Aether aether = Aether.builder().systemSession(anySession(), system).build();
+      VersionResolver resolver = new VersionResolver(aether);
+
+      resolver.resolveVersion("something", "something", "1.0");
+      fail();
+    } catch (VersionRangeResolutionException e) {
+      // This should be caught by the VersionResolver
+      fail();
+    } catch (ArtifactBuilder.InvalidArtifactCoordinateException e) {
+      String expected =
+          "Unable to find a version for something:something:1.0 due to Invalid Range Result";
+      assertThat(e.getMessage()).containsMatch(expected);
+    }
+  }
+
+  /**
+   * Asserts that given a soft pinned version specification, it selects that version,
+   * and does not get the highest version. "3.4" is an example of a soft pinned version specification.
+   */
+  //TODO(petros): implicitly using Maven central for resolving version. This is sketchy.
+  @Test
+  public void softPinnedVersions() {
+    try {
+      String version =
+          defaultResolver().resolveVersion("something", "something", "3.4");
```

Comment:
Thank you for pointing this out. It's quite unfortunate aether is not as complete as hoped for. That said, I have been playing around with a few ideas. For any given artifact `a:b:version-spec`, we can make a version range request for `a:b:[,)` and aether will return a list of all potential versions of that artifact. 

Here is one approach we could take. Whenever `resolveVersion` is called, we can simply make the request for `[,)`, i.e. the unbounded request. With all potential versions, we could then manually select the version from the available versions. This would push the brunt of selecting versions on us. However, it is much more reliable than what I have right now.